### PR TITLE
telemetry: get rid of the tracing_fmt layer and use our own optimized one

### PR DESCRIPTION
**Pull request by bnjjj**
Some perf improvements to lazy parse the attribute for logs

```
NEW ===
Summary:
  Total:        78.7140 secs
  Slowest:      0.1418 secs
  Fastest:      0.0055 secs
  Average:      0.0196 secs
  Requests/sec: 2540.8439


OLD ===
Summary:
  Total:        86.7263 secs
  Slowest:      0.1005 secs
  Fastest:      0.0065 secs
  Average:      0.0216 secs
  Requests/sec: 2306.1057
```

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by BrynCooke — APPROVED** [review]
looks great, but let's follow up with tests before merging to dev.

File: apollo-router/src/plugins/telemetry/fmt_layer.rs [hunk 1]
```diff
@@ -15,37 +28,198 @@ pub(crate) fn create_fmt_layer(
...
+                    .filter_map(|(k, v)| Some((k.into(), v.maybe_to_otel_value()?))),
+            );
+        } else {
+            eprintln!("cannot access to LogAttributes, this is a bug");
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        thread_local! {
+            static BUF: RefCell<String> = RefCell::new(String::new());
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/fmt_layer.rs:134]: nice

File: apollo-router/src/plugins/telemetry/fmt_layer.rs [hunk 2]
```diff
@@ -15,37 +28,198 @@ pub(crate) fn create_fmt_layer(
...
+    fn record_bool(&mut self, field: &Field, value: bool) {
+        self.values
+            .insert(field.name(), serde_json::Value::from(value));
+    }
+
+    /// Visit a string value.
+    fn record_str(&mut self, field: &Field, value: &str) {
+        let field_name = field.name();
+        match field_name {
+            "code.filepath" | "code.namespace" | "code.lineno" | "thread.id" | "thread.name" => {}
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/fmt_layer.rs:204]: does this translate to `field_name == "code.filepath" || field_name == "code.namespace" || etc`? We had an issue elsewhere where we checked if a span name was in a list, and that was surprisingly expensive. It was fixed by moving to a `HashSet`: https://github.com/apollographql/router/pull/3656/

File: apollo-router/src/plugins/telemetry/formatters/mod.rs [hunk 3]
```diff
@@ -75,7 +108,7 @@ pub(crate) fn filter_metric_events(event: &tracing::Event<'_>) -> bool {
     })
 }
 
-pub(crate) fn to_map(resource: Resource) -> BTreeMap<String, serde_json::Value> {
+pub(crate) fn to_map(resource: Resource) -> HashMap<String, serde_json::Value> {
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/formatters/mod.rs:111]: do we need the attributes to stay in the same order everywhere?

File: apollo-router/src/plugins/telemetry/formatters/json.rs [hunk 4]
```diff
@@ -93,101 +79,85 @@ where
...
+        // Get otel attributes
+        {
+            let otel_attributes = ext
+                .get::<OtelData>()
+                .and_then(|otel_data| otel_data.builder.attributes.as_ref());
+            if let Some(otel_attributes) = otel_attributes {
+                for (key, value) in otel_attributes.iter().filter(|(k, _)| {
+                    let key_name = k.as_str();
+                    !key_name.starts_with(APOLLO_PRIVATE_PREFIX)
+                        && !EXCLUDED_ATTRIBUTES.contains(&key_name)
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/formatters/json.rs:92]: as said ion another file: `EXCLUDED_ATTRIBUTES` should be a `HashSet`

File: apollo-router/src/plugins/telemetry/formatters/mod.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/formatters/mod.rs:111]: No I don't think it should be a requirement

**Review by garypen — COMMENTED** [review]
Nice performance improvement. Do we need any new tests?

File: apollo-router/src/plugins/telemetry/formatters/mod.rs [hunk 5]
```diff
@@ -11,13 +11,23 @@ use tracing::Subscriber;
...
+use tracing_subscriber::layer::Context;
 use tracing_subscriber::registry::LookupSpan;
 
 use crate::metrics::layer::METRIC_PREFIX_COUNTER;
 use crate::metrics::layer::METRIC_PREFIX_HISTOGRAM;
 use crate::metrics::layer::METRIC_PREFIX_MONOTONIC_COUNTER;
 use crate::metrics::layer::METRIC_PREFIX_VALUE;
 
+pub(crate) const APOLLO_PRIVATE_PREFIX: &str = "apollo_private.";
+pub(crate) const EXCLUDED_ATTRIBUTES: &[&str] = &[
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/formatters/mod.rs]: Added ! Thank you

File: apollo-router/src/plugins/telemetry/config_new/logging.rs [hunk 6]
```diff
@@ -328,6 +328,8 @@ impl Default for JsonFormat {
 #[derive(Deserialize, JsonSchema, Clone, Debug, Eq, PartialEq)]
 #[serde(deny_unknown_fields, rename_all = "snake_case", default)]
 pub(crate) struct TextFormat {
+    /// Enable ansi (default: true)
+    pub(crate) ansi: bool,
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/config_new/logging.rs:332]: @BrynCooke I know we previously had that configuration in our first configuration draft. Do you have any opinions ?

**Comment by bnjjj** [comment]
@garypen Yes I will add tests now, but in a follow up PR to make sure PRs stay small and easy to review. But it will be easier to test our logs now we have our own layer I think

File: apollo-router/src/plugins/telemetry/config_new/logging.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/config_new/logging.rs:332]: Let's change it to `ansi_escape_codes` and update the comment on the rust doc.

**Pull request merged**